cmd/disk-cache.go

```diff
@@ -0,0 +1,894 @@
+		return putObjectPartFn(bucket, object, uploadID, partID, data)
+	}
+
+	info = PartInfo{}
+	// Initialize pipe to stream data to backend
+	pipeReader, pipeWriter := io.Pipe()
+	hashReader, err := hash.NewReader(pipeReader, size, data.MD5HexString(), data.SHA256HexString())
+	if err != nil {
+		return
+	}
+	// Initialize pipe to stream data to cache
+	rPipe, wPipe := io.Pipe()
+	cHashReader, err := hash.NewReader(rPipe, size, data.MD5HexString(), data.SHA256HexString())
+	if err != nil {
+		return
+	}
+	pinfoCh := make(chan PartInfo)
+	go func() {
+		info, err = putObjectPartFn(bucket, object, uploadID, partID, hashReader)
+		if err != nil {
+			close(pinfoCh)
+			pipeWriter.CloseWithError(err)
+			return
+		}
+		pinfoCh <- info
+	}()
+	go func() {
+		if _, perr := dcache.PutObjectPart(bucket, object, uploadID, partID, cHashReader); perr != nil {
+			wPipe.CloseWithError(perr)
+			return
+		}
+	}()
+
+	mwriter := io.MultiWriter(pipeWriter, wPipe)
+	_, err = io.Copy(mwriter, data)
+	if err != nil {
+		pipeWriter.CloseWithError(err)
+		wPipe.CloseWithError(err)
+		return PartInfo{}, err
+	}
+	info = <-pinfoCh
+	return info, err
+}
+
+// AbortMultipartUpload - aborts multipart upload on backend and cache.
+func (c cacheObjects) AbortMultipartUpload(bucket, object, uploadID string) error {
+
+	abortMultipartUploadFn := c.AbortMultipartUploadFn
+
+	if c.isCacheExclude(bucket, object) {
+		return abortMultipartUploadFn(bucket, object, uploadID)
+	}
+
+	dcache, err := c.cache.getCacheFS(bucket, object)
+	if err != nil {
+		// disk cache could not be located,execute backend call.
+		return abortMultipartUploadFn(bucket, object, uploadID)
+	}
+	// execute backend operation
+	err = abortMultipartUploadFn(bucket, object, uploadID)
+	if err != nil {
+		return err
+	}
+	// abort multipart upload on cache
+	dcache.AbortMultipartUpload(bucket, object, uploadID)
+	return nil
+}
+
+// CompleteMultipartUpload - completes multipart upload operation on backend and cache.
+func (c cacheObjects) CompleteMultipartUpload(bucket, object, uploadID string, uploadedParts []CompletePart) (objInfo ObjectInfo, err error) {
+
+	completeMultipartUploadFn := c.CompleteMultipartUploadFn
+
+	if c.isCacheExclude(bucket, object) {
+		return completeMultipartUploadFn(bucket, object, uploadID, uploadedParts)
+	}
+
+	dcache, err := c.cache.getCacheFS(bucket, object)
+	if err != nil {
+		// disk cache could not be located,execute backend call.
+		return completeMultipartUploadFn(bucket, object, uploadID, uploadedParts)
+	}
+	// perform backend operation
+	objInfo, err = completeMultipartUploadFn(bucket, object, uploadID, uploadedParts)
+	if err != nil {
+		return
+	}
+	// create new multipart upload in cache with same uploadID
+	dcache.CompleteMultipartUpload(bucket, object, uploadID, uploadedParts)
+	return
+}
+
+// initCaches initializes the cacheFSObjects for the "drives" specified in config.json
+// or the global env overrides.
+func initCaches(c CacheConfig) (*diskCache, error) {
```

Comment:
we should call this `newCache` - `init` is not meaningful since this returns values..